articles/actions.md

```diff
@@ -0,0 +1,338 @@
+
+Python:
+
+- [examples/rclpy/actions](https://github.com/ros2/examples/tree/actions_proposal/rclpy/actions)
+
+## Middleware implementation
+
+Under the hood, an action is made up of three services and two topics, each descibed in detail here.
+
+![Action Client/Server Interaction Overview](../img/actions/interaction_overview.png)
+
+### Goal Submission Service
+
+* **Direction**: Client calls Server
+* **Request**: Description of goal
+* **Response**: Whether goal was accepted or rejected, a unique identifier for the goal, and the time when the goal was accepted.
+
+The purpose of this service is to submit a goal to the action server.
+It is the first service called to begin an action, and is expected to return quickly.
+A user-define description of the goal is sent as the request.
+The response indicates whether or not the goal was accepted, and if so the identifier the server will use to describe the goal.
+
+The QoS settings of this service should be set the so the client is guaranteed to receive a response or an action could be executed without a client being aware of it.
+
+### Cancel Request Service
+
+* **Direction**: Client calls Server
+* **Request**: Goal identifier, time stamp
+* **Response**: Goals that will be attempted to be canceled
+
+The purpose of this service is to request to cancel one or more goals on the action server.
+A cancellation request may cancel multiple goals.
+The result indicates which goals will be attempted to be canceled.
+Whether or not a goal is actually canceled is indicated by the status topic and the result service.
+
+The cancel request policy is the same as in ROS 1.
+
+* If the goal ID is empty and time is zero, cancel all goals
+* If the goal ID is empty and time is not zero, cancel all goals accepted at or before the time stamp
+* If the goal ID is not empty and time is not zero, cancel the goal with the given id regardless of the time it was accepted
+* If the goal ID is not empty and time is zero, cancel the goal with the given id and all goals accepted at or before the time stamp
+
+### Get Result Service
+
+* **Direction**: Client calls Server
+* **Request**: Goal ID
+* **Response**: Status of goal and user defined result
+
+The purpose of this service is to get the final result of a service.
+After a goal has been accepted the client should call this service to receive the result.
+The result will indicate the final status of the goal and any user defined data.
+
+Once the server sends the result to the client it should free up any resources used by the action.
+If the client never asks for the result then the server should discard the result after a timeout period.
+
+### Goal Status Topic
+
+* **Direction**: Server publishes
+* **Content**: List of in-progress goals with: Goal ID, time accepted, and an enum indicating the status
+
+This topic is published by the server to broadcast the status of goals it has accepted.
```

Comment:
> Same problem as feedback (https://github.com/ros2/design/pull/193/files#r225264321) with clients receiving data they don't care about.

I don't think a client will subscribe to this topic. It is meant for introspection only.

>  Additionally, there are issues where a 'lost on the wire' status message could cause a failure to transition properly client side.
>
> I think it's really important to consider how this respin of actions is going to work on unreliable networks, as that's one of the major applications (and goals?) of ROS2/DDS.

Agreed. The QoS settings definitely need to be exposed to the user.

> Reading over this and the 'get_result' section above, would it make sense to have the action server 'push' results and statuses to clients via a service? It would be up to the client to specify a callback where to process the results for any submitted goals.

Since ROS 2 services are asynchronous (in `rcl` and `rclpy`, exposing this in `rclcpp` is tracked in ros2/rclcpp#491) I think the result is the same. Calling `get_result` is like asking the server to push the result to the action client in the future. It just happens the action result is communicated as the response to a service call.

Reply:
> If the result is the same, then unless I'm missing something, the point of having the client call a service to retrieve the result, is to avoid having the result proliferated to every client via topic, otherwise the 'request' portion of get_result seems redundant.

That case wasn't considered. The assumption is that there are usually only a handful of clients (in most cases just one) per server. Are you currently using many action clients per server in ROS 1? If so, how well does it perform?

Services make some parts of the client implementation simpler. The client doesn't need to track the server's state machine. `wait_for_action_server` could reuse `wait_for_service`. It does have a drawback that for clients who never fetch the result, the server will have to decide how long to hold onto it before discarding it.

Reply:
Starting to get our 'results' mixed up :)

I've seen low-to-mid double digits clients per server. To some degree that's probably because of how inflexible 'services' were in ROS1, but regardless...

I don't want to advise for over-optimization, and I appreciate that with a central `rcl` implementation of action comms, you have the option to restructure 'later', but it may be worthwhile to make certain scaling-friendly decisions up front. Transmitting N results/feedbacks to M clients over a potentially limited comms channel feels like one of those decisions. If `get_result` is an acceptable strategy, then maybe `get_feedback` is as well.